### PR TITLE
Ruby 2.7.1 warnings: AHORA SI QUE SI, CABESA

### PR DESCRIPTION
**Pull request by julitrows**
This PR addresses the final warnings we were getting related to Ruby 2.7 change in argument delegation.

As it is, all specs are green in both Ruby 2.6 and 2.7.1

Suggestions to beautify the `split_args` and `build_instance` are welcome

### CODE REVIEW

File: spec/injectable_spec.rb [hunk 1]
```diff
@@ -159,6 +159,38 @@ def call
     end
   end
 
+  context 'with dependencies that have a with: keyword option and other with array' do
+    before do
+      DepWithBothArgs = Class.new do
+        attr_reader :somearg, :options
+        def initialize(somearg, **options)
```
Comment (rewritten) [inline spec/injectable_spec.rb:166]: I think it would fail if the first argument is a hash, as it would get wrongly merged into kwargs. I'll take a look tomorrow.

File: lib/injectable/dependency.rb [hunk 2]
```diff
@@ -2,14 +2,30 @@ module Injectable
...
 
     private
 
-    def wrap_args(args)
+    def split_args(args)
+      positional_args = []
+      kwargs = {}
+
+      preprocess_args(args).each do |arg|
+        arg.is_a?(Hash) ? kwargs.merge!(arg) : positional_args << arg
```
Comment (rewritten) [inline lib/injectable/dependency.rb:17]: It probably makes sense to check ActiveSupport implementation of `extract_options` which has exactly the same purpose.

**Review by rewritten — COMMENTED** [review]
Not sure about this, I would like to add more test cases.

**Comment by iovis** [comment]
> Not sure about this, I would like to add more test cases.

Yeah, please add/suggest as many test cases as necessary, this is a first draft to have a starting point

**Comment by iovis** [comment]
I switched to Draft till we iron out any possible edge cases

File: lib/injectable/dependency.rb (hunk 2 above)
Comment (iovis) [reply lib/injectable/dependency.rb:17]: Seems like it just checks if the last element is a hash

```rb
# File activesupport/lib/active_support/core_ext/array/extract_options.rb, line 24
  def extract_options!
    if last.is_a?(Hash) && last.extractable_options?
      pop
    else
      {}
    end
  end

# File activesupport/lib/active_support/core_ext/array/extract_options.rb, line 9
  def extractable_options?
    instance_of?(Hash)
  end
```

File: spec/injectable_spec.rb (hunk 1 above)
Comment (jantequera) [reply spec/injectable_spec.rb:166]: If you can provide a breaking spec with that we can work on the solution together. As David says, bring in as many edge cases as you can think of :D

**Comment by rewritten** [comment]
@jantequera @iovis now it covers every coverable case I can come up with. 

Of course if the user passes a symbolic hash as the last positional argument, and the dependency has explicit kwarg initializer incompatible with the passed hash, it will explode, but I don't think it should be covered.

This is an example of the failing case:

```rb
class MyDependency
  def initialize(positional = nil, kw1: nil); end
end

class MyInjectable
  include Injectable

  dependency :my_dependency, with: [{foo: :bar}]
end
```

`:foo` will be passed as kwarg to the initializer, and it will raise an exception. Tell me if you want to cover also this case, because it will make the code a lot more complex, as I will have to introspect the signature of the initializer.

**Comment by iovis** [comment]
Seems like there are still some warnings for ruby 2.7
Screen Shot 2021-03-15 at 12 56 33

**Comment by rewritten** [comment]
> Seems like there are still some warnings for ruby 2.7


@iovis  These are unrelated to ruby 2.7, it complains for a constant defined in a block, and we should actually refactor the code to have those constants defined only once.

**Comment by julitrows** [comment]
@iovis where did you get those? Never saw them running the specs locally 🤔

**Comment by iovis** [comment]
In the PR checks for the ruby 2.7 build

**Review by iovis — APPROVED** [review]
LGTM

**Comment by iovis** [comment]
> Of course if the user passes a symbolic hash as the last positional argument, and the dependency has explicit kwarg initializer incompatible with the passed hash, it will explode, but I don't think it should be covered.

We can revisit if we run into this
